Partial GCC command option support

0x59616e opened the pull request:
Before this change, we have to redirect the stdin 
and stdout :

              ./mzcc < input.c > output.S

which is not intuitive, and not what we are used to.
After this change, we can use MazuCC just like gcc :

              ./mzcc -o output.S input.c

Commit:
Make the user interface to be more like gcc.

After this change, the user does not need to redirect the stdin and
stdout. They can use MazuCC just like gcc. To be more accurately,
before this change, we have to redirect the stdin and stdout :

              ./mzcc < input.c > output.S

which is not intuitive, and not what we are used to.
After this change, we can use MazuCC just like gcc :

              ./mzcc -o output.S input.c

jserv commented on main.c:
```diff
@@ -1,11 +1,73 @@
+}
+static void parseopt(int argc, char **argv)
+{
+    if (argc == 1) {
```
I would expect the following changes for consistent style:
```c
    if (argc < 2) {
        usage();
        return;
    }
    int opt;
    while ((opt = getopt(argc, argv, "o:d:")) != -1) {
        ...
```

jserv commented on codegen_x64.c:
```diff
@@ -5,7 +5,7 @@
 static int TAB = 8;
 static List *functions = &EMPTY_LIST;
-
+FILE *outputfp;
```
How about changing it to `outfp`?

jserv commented on main.c:
```diff
@@ -1,11 +1,73 @@
+        usage();
+    } else {
+        int opt;
+        while ((opt = getopt(argc, argv, "o:d:")) != -1) {
```
The `getopt()`  function is provided by POSIX.1-2001, POSIX.1-2008, and POSIX.2 with environment variable `POSIXLY_CORRECT`. I wish to avoid the use of `getopt` for such simple cases. You might check how AMaCC manipulates the arguments: https://github.com/jserv/amacc

0x59616e commented:
Thanks for your advice. I appreciate that !
I'll make the changes you suggested. Thank you.

Commit:
Eliminate the use of getopt() function

jserv commented:
> Thanks for your advice. I appreciate that !
> I'll make the changes you suggested. Thank you.

Thanks for your contribution. I saw the style changes. Please submit another git commit to eliminate the use of `getopt`.

0x59616e commented:
Thanks for your reply. I'm working on this. Thank you for your patience !

jserv commented:
In addition, file `README.md` should be updated accordingly for the sake of command line adaption.

0x59616e commented:
OK, I'll make the changes after all the work is finished. Thank you.

0x59616e commented:
There is a minor error in READMD.md. I'll fix that ASAP. Thank you.

jserv commented:
The existing build targets should be verified as well. Please ensure the correctness of `make check`.

0x59616e commented:
Sorry, I didn't notice that. I'll fix it as soon as possible. Thanks for your patience.

jserv commented on main.c:
```diff
@@ -1,11 +1,97 @@
+static FILE *open_output_file()
+{
+    FILE *tmp = stdout;
+    if (outfile != NULL && ((tmp = fopen(outfile, "w")) == NULL)) {
```
Change this line to
```c
if (outfile && ((tmp = fopen(outfile, "w"))) {
```

0x59616e replied:
OK, I'll make that change. Thanks

jserv commented on main.c:
```diff
@@ -1,11 +1,97 @@
+                outfile = *argv;
+                break;
+            case 'd':
+                if (strcmp(*argv, "-dump-ast")) {
```
Change to
```c
if (!strcmp(*argv, "-dump-ast")) {
    dump_ast = true;
    break;
}
```

jserv commented on main.c:
```diff
@@ -1,11 +1,97 @@
+
+            "  -o filename     Output to the specified file\n"
+            "  -dump-ast       Print Abstract Syntax Tree\n\n");
+}
```
Append new line.

jserv commented on main.c:
```diff
@@ -1,11 +1,97 @@
+{
+    usage();
+    exit(1);
+}
```
Ditto. New line.

jserv commented on main.c:
```diff
@@ -1,11 +1,97 @@
+                print_usage_and_exit();
+            }
+        } else {
+            if (infile == NULL) {
```
Change to
```c
    if (!infile) {
        // The second non-option argument is not what we expect.
        print_usage_and_exit();
    }
    infile = argv[0];
```

jserv commented on main.c:
```diff
@@ -1,11 +1,97 @@
+            }
+        }
+    }
+}
```
Ditto. New line.

jserv commented on main.c:
```diff
@@ -1,11 +1,97 @@
+        exit(1);
+    }
+    return tmp;
+}
```
Ditto. New line.

jserv commented on main.c:
```diff
@@ -1,11 +1,97 @@
+        exit(1);
+    }
+    return tmp;
+}
```
Ditto. New line.

0x59616e replied:
Do you mean appending a new line after the right braces ?

jserv replied:
Exactly.

jserv commented:
Use `clang-format -i` for consistent coding style.

jserv commented on main.c:
```diff
@@ -1,11 +1,97 @@
+}
+static FILE *open_input_file()
+{
+    if (infile == NULL) {
```
Don't use `infile == NULL`. Replace it with `!infile`.

jserv commented on main.c:
```diff
@@ -1,11 +1,97 @@
+    usage();
+    exit(1);
+}
+static void parseopt(int argc, char **argv)
```
Rename the function to `parse_args`.

jserv commented on README.md:
````diff
@@ -47,8 +47,19 @@ Q . . . . . . .
+} a;
+```
 ```shell
-echo 'struct {int x; char y; struct { int t; } z; } a;' | ./mzcc --dump-ast
````
In gcc, specifying that a filename is `-`, the standard convention for saying 'standard input'.
Thus, both of the following lines are legal to generate executables:
```bash
echo 'int main() { return 0; }' | gcc -xc -O -Wall -o app /dev/stdin
echo 'int main() { return 0; }' | gcc -xc -O -Wall -o app -
```

It would be great if you can support the feature.

0x59616e replied:
That looks good. I'll try that. Thanks

jserv replied:
Accepting source listing from stdin is important to implement test suite for verifying compilers. That is, we don't have to prepare lots of source files in advance.

Commit:
Support the standard input option.

Before this commit, we must to specify a input file in the command line.
After this commit, we can use option '-' to use standard input.

jserv commented on Makefile:
```diff
@@ -51,15 +51,15 @@ check: nqueen $(TESTS)
 
 tests/%.s: tests/%.c $(TARGET)
-	./mzcc < $< > $@
+	./mzcc < $< > $@ -
```
Why didn't you use the new option `-o`?

jserv commented on Makefile:
```diff
@@ -51,15 +51,15 @@ check: nqueen $(TESTS)
 nqueen: sample/nqueen.c $(TARGET)
 	$(VECHO) "  MazuCC\t$<\n"
-	$(Q)./mzcc < $< > ${<:.c=.s}
+	$(Q)./mzcc < $< > ${<:.c=.s} -
```
Ditto. Use `-o` option.

jserv commented on main.c:
```diff
@@ -1,11 +1,104 @@
+        "\n"
+
+        "  -o filename            Output to the specified file\n"
+        "  -dump-ast              Print Abstract Syntax Tree\n"
```
It should be `--dump-ast` or `-d` for short.

jserv commented:
Use `git rebase -i` to rework the git commits with minimizing changes.

jserv commented on Makefile:
```diff
@@ -51,15 +51,15 @@ check: nqueen $(TESTS)
 
 tests/%.s: tests/%.c $(TARGET)
-	./mzcc < $< > $@
+	./mzcc < $< -o $@ -
```
Rewrite as `./mzcc -o $@ $<`

jserv commented on Makefile:
```diff
@@ -51,15 +51,15 @@ check: nqueen $(TESTS)
 nqueen: sample/nqueen.c $(TARGET)
 	$(VECHO) "  MazuCC\t$<\n"
-	$(Q)./mzcc < $< > ${<:.c=.s} -
+	$(Q)./mzcc < $< -o ${<:.c=.s} -
```
Ditto.

0x59616e commented:
I'm trying to rebase the commits. Thanks for your patience.

0x59616e commented:
When I wanted to push, I had no choice but to use the -f option. I don't know if this is ok.

jserv commented:
> When I wanted to push, I had no choice but to use the -f option. I don't know if this is ok.

It is an expected operation.

jserv commented on Makefile:
```diff
@@ -51,15 +51,15 @@ check: nqueen $(TESTS)
 
 tests/%.s: tests/%.c $(TARGET)
-	./mzcc < $< > $@
+	./mzcc -o $@ < $< - 
```
Don't abuse `-`. Keep the command line as similar as gcc does.

0x59616e replied:
Ohh...I was out of my mind Orz

jserv commented on main.c:
```diff
@@ -1,11 +1,103 @@
+            "Usage : mzcc [ input file | - ]\n"
+            "\n"
+            "  -o filename            Output to the specified file\n"
+            "  -d --dump-ast          Print Abstract Syntax Tree\n"
```
Should be `-d or --dump-ast`

0x59616e replied:
OK, I'll remove the -d option

jserv commented on Makefile:
```diff
@@ -51,15 +51,15 @@ check: nqueen $(TESTS)
 
 tests/%.s: tests/%.c $(TARGET)
-	./mzcc < $< > $@
+	./mzcc -o $@ $<  
```
Remove trailing space.

jserv commented on main.c:
```diff
@@ -1,11 +1,103 @@
 #include <stdbool.h>
 #include <stdio.h>
 #include <string.h>
+#include <unistd.h>
```
This inclusion can be eliminated.

jserv commented on main.c:
```diff
@@ -1,11 +1,103 @@
 {
-    bool dump_ast = ((argc > 1) && !strcmp(argv[1], "--dump-ast"));
+    parse_args(argc, argv);
+    stdin = open_input_file();
```
Avoid using variable name "stdin", which might cause confusion.

0x59616e replied:
I change the function like this :

```c
static void open_input_file()
{
    if (!infile) {
        printf("Input file is not specified\n\n");
        print_usage_and_exit();
    }
    
    if (!freopen(infile, "r", stdin)) {
        printf("Can not open file %s\n", infile);
        exit(1);
    }
}
```
Instead of returning a pointer, we redirect the stdin in this function.
Is that better ?

jserv commented on README.md:
````diff
@@ -21,13 +21,13 @@ MazuCC is known to work on both GNU/Linux and macOS.
 Use MazuCC to compile C source:
 ```shell
-$ ./mzcc < sample/nqueen.c
+$ ./mzcc  sample/nqueen.c
````
Mention the alternative way as following:
> Alternatively, MazuCC accepts the stream from standard input. The equivalent form for the above command is:
> ```shell
> cat sample/nqueen.c | ./mzcc -
> ```

jserv commented on main.c:
```diff
@@ -1,11 +1,103 @@
+#include <unistd.h>
 #include "mzcc.h"
 
+char *outfile = NULL, *infile = NULL;
```
`infile` should be declared as a static variable since it would not be exposed.

Commit:
Update the test script and README.md

Update the test script and READMD.md according to the changes on
the interface.

Commit:
Rewrite the open_input_file function

We now use freopen instead of fopen to redirect the stdin.

Also rewrite the open_output_file function to keep the style
consistent.

jserv commented on README.md:
````diff
@@ -21,13 +21,20 @@ MazuCC is known to work on both GNU/Linux and macOS.
+ form for the above command is:
+
+ ```shell
+ cat sample/nqueen.c | ./mzcc -
````
Add leading `$` for consistency.

jserv commented on codegen_x64.c:
```diff
@@ -5,7 +5,7 @@
 static int TAB = 8;
 static List *functions = &EMPTY_LIST;
-
+FILE *outfp;
```
Add comment like this:
```C
/* FIXME: main program should take extern variables from codegen. */
```

jserv commented:
Delete the generated file `hmygod`.

jserv commented on main.c:
```diff
@@ -3,9 +3,96 @@
+static void usage()
+{
+    fprintf(
+        stdout,
```
Rewrite as following:
> mzcc [options] filename
> OPTIONS
>     -o <file> : Write output to the specified file.
>     --dump-ast : Dump abstract syntax tree (AST)

jserv commented on codegen_x64.c:
```diff
@@ -5,6 +5,7 @@
 static char *REGS[] = {"rdi", "rsi", "rdx", "rcx", "r8", "r9"};
 static int TAB = 8;
 static List *functions = &EMPTY_LIST;
+/* FIXME: main program should take extern variables from codegen. */
```
Insert one blank line before FIXME comment.

jserv commented on main.c:
```diff
@@ -3,9 +3,93 @@
+            "  --dump-ast             Dump abstract syntax tree(AST)\n");
+}
+
+void print_usage_and_exit()
```
At present, function `print_usage_and_exit` is not exposed. So, mark it as `static` as well.

Commit:
Rewrite the usage message, add FIXME in codegen_x64.c

Commit:
Mark print_usage_and_exit as static function

Since print_usage_and_exit function is not exposed, so we delcare it
as a static function.

jserv commented:
Thank @a9502854519 again for contributing to MazuCC.

0x59616e commented:
Great great thanks for your instruct and patience. It's my honor to contribute to this project. Wish you have a happy lunar new year :)
